app/src/main/java/com/mapbox/services/android/navigation/testapp/example/ui/ExampleViewModel.kt

```diff
@@ -0,0 +1,143 @@
+  private val routeFinder: ExampleRouteFinder
+  private val accessToken: String = instance.resources.getString(R.string.mapbox_access_token)
+
+  init {
+    // Initialize the location engine
+    val locationEngineProvider = LocationEngineProvider(getApplication())
+    locationEngine = locationEngineProvider.obtainBestLocationEngineAvailable()
+    locationEngineListener = ExampleLocationEngineListener(locationEngine, location)
+    locationEngine.addLocationEngineListener(locationEngineListener)
+    locationEngine.priority = LocationEnginePriority.HIGH_ACCURACY
+    locationEngine.fastestInterval = ONE_SECOND_INTERVAL
+
+    // Initialize the speech player and pass to milestone event listener for instructions
+    val english = US.language // TODO localization
+    val speechPlayerProvider = SpeechPlayerProvider(getApplication(), english, true, accessToken)
+    speechPlayer = NavigationSpeechPlayer(speechPlayerProvider)
+
+    // Initialize navigation and pass the LocationEngine
+    navigation = MapboxNavigation(getApplication(), accessToken)
+    navigation.locationEngine = locationEngine
+    navigation.addMilestoneEventListener(ExampleMilestoneEventListener(milestone, speechPlayer))
+    navigation.addProgressChangeListener(ExampleProgressChangeListener(location, progress))
+    navigation.addOffRouteListener(ExampleOffRouteListener(this))
+
+    // For fetching new routes
+    routeFinder = ExampleRouteFinder(routes, accessToken)
+  }
+
+  fun activateLocationEngine() {
+    locationEngine.activate()
+  }
+
+  fun findRouteToDestination() {
+    location.value?.let { location ->
+      destination.value?.let { destination ->
+        routeFinder.findRoute(location, destination)
+      }
+    }
+  }
+
+  fun updatePrimaryRoute(primaryRoute: DirectionsRoute) {
+    routeFinder.primaryRoute = primaryRoute
+  }
+
+  fun canNavigate(): Boolean {
+    return routeFinder.primaryRoute != null
+  }
+
+  fun startNavigation() {
+    routeFinder.primaryRoute?.let {
+      navigation.startNavigation(it)
+      removeLocationEngineListener()
+    }
+  }
+
+  fun stopNavigation() {
+    addLocationEngineListener()
+    navigation.stopNavigation()
+  }
+
+  fun retrieveNavigation(): MapboxNavigation {
+    return navigation
+  }
+
+  fun reverseGeocode(point: LatLng) {
+    val reverseGeocode = MapboxGeocoding.builder()
+            .accessToken(Mapbox.getAccessToken()!!)
+            .query(Point.fromLngLat(point.longitude, point.latitude))
+            .geocodingTypes(GeocodingCriteria.TYPE_ADDRESS)
+            .build()
+    reverseGeocode.enqueueCall(object : Callback<GeocodingResponse> {
+      override fun onResponse(call: Call<GeocodingResponse>, response: Response<GeocodingResponse>) {
+        geocode.value = response.body()
+      }
+
+      override fun onFailure(call: Call<GeocodingResponse>, throwable: Throwable) {
+        Timber.e(throwable, "Geocoding request failed")
+      }
+    })
+  }
+
+  fun onDestroy() {
```

Comment:
`AndroidViewModel` has an `onCleared` method that we can override. It's called when the `Activity` is about to be destroyed so it'd remove the necessity of having to call this function from the `ExampleActivity#onDestroy` > `ExamplePresenter#onDestroy`